### PR TITLE
chore(tests): Build examples as matrix - 1 build per example

### PR DESCRIPTION
**Pull request by skorfmann**
This PR is a first step towards drastically reduced build times - Related to #405 

It tries the concept of building a dynamic build matrix for each example, in order to parallelize these builds

### CODE REVIEW

File: tools/build-examples-sequential.sh [hunk 1]
```diff
@@ -9,7 +9,7 @@ export CI='1'
 
 BUILD_TARGET=$1
 
-npx lerna run --stream --concurrency 1 --no-bail --scope @examples/${BUILD_TARGET}* reinstall
-npx lerna run --stream --concurrency 1 --no-bail --scope @examples/${BUILD_TARGET}* build
-npx lerna run --stream --concurrency 1 --no-bail --scope @examples/${BUILD_TARGET}* synth
+npx lerna run --stream --concurrency 1 --no-bail --scope ${BUILD_TARGET}* reinstall
+npx lerna run --stream --concurrency 1 --no-bail --scope ${BUILD_TARGET}* build
+npx lerna run --stream --concurrency 1 --no-bail --scope ${BUILD_TARGET}* synth
```
Comment (ansgarm) [inline tools/build-examples-sequential.sh:15]: If we now target only one test in the `BUILD_TARGET` we can remove this script, can't we? Or does you change only apply to PR builds and we still need this elsewhere (e.g. when running builds on `main`)?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -51,15 +51,28 @@ jobs:
...
+
   linux_integration:
+    needs: build-test-matrix
     runs-on: ubuntu-latest
     strategy:
-      matrix:
-        terraform: ["0.14.11", "0.15.4"]
-        target: ["typescript", "python", "java", "csharp", "go"]
+      fail-fast: false
+      matrix: ${{fromJSON(needs.build-test-matrix.outputs.examples)}}
```
Comment (jsteinich) [inline .github/workflows/build.yml:73]: Should this be ...outputs.tests instead of ...outputs.examples?

Comment (skorfmann) [reply .github/workflows/build.yml:74]: Yes, thanks!

File: tools/build-examples-sequential.sh (hunk 1 above)
Comment (skorfmann) [reply tools/build-examples-sequential.sh]: Yep, cleaned it up a bit

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -76,43 +65,66 @@ jobs:
...
+          tests=$(cd test && npx jest --listTests | jq  -R -s --arg cwd "$(pwd)/" -c 'split("\n") | map(select(length > 0)) | map(sub($cwd; "")) | { target: values, terraform: ["0.14.11", "0.15.4"]}')
+          echo $tests
+          echo "::set-output name=tests::$tests"
+
+  linux_integration:
+    needs: prepare-integration-tests
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix: ${{fromJSON(needs.prepare-integration-tests.outputs.tests)}}
```
Comment (DanielMSchmidt) [inline .github/workflows/build.yml:93]: fancy 🚀

**Review by DanielMSchmidt — APPROVED** [review]

File: examples/typescript/backends/remote/package.json [hunk 4]
```diff
@@ -1,10 +1,11 @@
 {
-  "name": "@deactivated-examples/typescript-backends-remote",
+  "//": "This example test is disabled via the 'private' attribute since it needs credentials (I assume at least)",
```
Comment (jsteinich) [inline examples/typescript/backends/remote/package.json:2]: Actually using it would need credentials, but just synthing doesn't need anything.

**Review by jsteinich — COMMENTED** [review]
Worth noting that there is functionality (aws provider gen, module gen) that is now only tested in examples for non-typescript languages. This isn't a problem now, but should be thought about if example tests are ever removed / moved out of PR flow.

**Pull request merged**